File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.cc
/******************************************************************************

                              Online C++ Compiler.
               Code, Compile, Run and Debug C++ program online.
Write your code in this editor and press "Run" button to compile and execute it.

*******************************************************************************/

#include <bits/stdc++.h>

using namespace std;

#define MAX 100001
#define INF 987654321
int _visit[MAX];
int dist[MAX];

int main()
{
  int n,k;
  cin >> n >> k;
  fill(dist,dist+MAX,INF);
  queue<int> q;
    
    dist[n]=0;
    q.push(n);
 

    while(!q.empty())
    {
        int now=q.front();
        q.pop();
        
        
        if(_visit[now]==1) continue;
        if(now==k) break;
        _visit[now]=1;
        
        if(now*2<=MAX-1&&dist[now*2]>dist[now])
        {
            q.push(now*2);
            dist[now*2]=dist[now];
            
        }
        
        if(now-1>=0&&dist[now-1]>dist[now]+1)
        {

            q.push(now-1);
            dist[now-1]=dist[now]+1;
        }
        if(now+1<=MAX-1&&dist[now+1]>dist[now]+1)
        {
            q.push(now+1);
            dist[now+1]=dist[now]+1;
        }
        
        
    }
    
    cout << dist[k] <<endl;
    
    
  
}

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.kt
import java.util.*
fun main(args: Array<String>) {
    val INF=987654321
    val MAX= 100001
    val (N,K)=readLine()!!.split(" ").map{it.toInt()}

    var visit:IntArray=IntArray(MAX,{0})
    var dist:IntArray=IntArray(MAX,{INF})
    var q: Queue<Int> = LinkedList<Int>()
    
    
    q.add(N)
    dist[N]=0
    
    while(q.peek()!=null)
    {
    	val now:Int=q.poll()
      if(now==K) break
      if(visit[now]==1) continue
      visit[now]=1

      var next:Int=2*now
      if(next<=MAX-1&&dist[next]>dist[now])
      {
        q.add(next)
        dist[next]=dist[now]
      }
      next=now+1
      if(next<=MAX-1&&dist[next]>dist[now]+1)
      {
        q.add(next)
        dist[next]=dist[now]+1
      }
      next=now-1
      if(0<=next&&dist[next]>dist[now]+1)
      {
        q.add(next)
        dist[next]=dist[now]+1
      }
    
    }
  	print(dist[K])
}